Use user scope for `eval()` invocation in `@artifact_str`

staticfloat opened the pull request:
This causes the eval to be invoked in user scope, rather than in the `Artifacts` module itself.

X-ref: https://github.com/JuliaLang/julia/pull/46755#pullrequestreview-1135759752

vchuravy commented on stdlib/Artifacts/src/Artifacts.jl:
```diff
@@ -662,7 +662,7 @@ macro artifact_str(name, platform=nothing, artifacts_toml_path=nothing)
         find_artifacts_toml(srcfile)
     else
-        eval(artifacts_toml_path)
+        Core.eval(__module__, artifacts_toml_path)
```
Instead of `eval` use `getfield`? But that may just be personal taste.

staticfloat replied:
How would you use `getfield`?  In this case, `artifacts_toml_path` could be an expression, rather than a quoted string.

vchuravy replied:
I would prefer not using a runtime/order dependent lookup here. `eval` inside a macro is for me a code-smell. Either the data is a literal and can be relied upon, or it is a runtime value and we should defer it's evaluation to inside the `quote` block.

simeonschaub commented:
Couldn't we just require this to always be a literal string? Consumers could still do `@eval @artifact_str "foo" $bar` if this was really needed. Calling `eval` in a macro just seems like a bad idea and I could see this causing issues in environments like Pluto.

KristofferC commented:
> Couldn't we just require this to always be a literal string?

That was the original idea but Elliot added the eval in (https://github.com/JuliaLang/julia/pull/46755/commits/0f309bc9faa12cf12e1a1d9b5dc06259ed129fea), likely due to some use case that I had not thought about.

StefanKarpinski commented:
Can we revert that? This seems quite fishy.

staticfloat committed:
[BinaryPlatforms] Change "shortest match" algorithm to "best match"

My assertion in the previous attempt to fix this issue was incorrect:

> We define a simpler match as one that has fewer tags overall.
> As these candidate matches have already been filtered to match the
> given platform, the only other tags that exist are ones that are in
> addition to the tags declared by the platform.  Hence, selecting the
> minimum in number of tags is equivalent to selecting the closest match.

This is demonstrably false, by my own test case:

```
platforms = Dict(
    Platform("x86_64", "linux") => "bad",
    Platform("x86_64", "linux"; sanitize="memory") => "good",
)
select_platform(platforms, Platform("x86_64", "linux"; sanitize="memory")) == "good"
```

In this case, because there exists a candidate that is _more general_
than the provided platform type, the shortest match is no longer the
best match.

This PR performs a more rigorous matching that works more reliably in
all cases.
